src/properties.ts

```diff
@@ -0,0 +1,311 @@
+   *   ],
+   * },
+   */
+  FILES = 'files',
+}
+
+export interface StoryPropertyBase<T> {
+  type: PropertyTypes;
+
+  /**
+   * label to display next to the field editor
+   * by default uses the property name itself
+   */
+
+  label?: string;
+
+  /**
+   * a default value for the property
+   */
+  value?: T;
+
+  /**
+   * hide the label from the property editor
+   */
+  hideLabel?: boolean;
+  /**
+   * hide the property editor for this property
+   * will only use the value
+   */
+
+  hidden?: boolean;
+  /**
+   * allows grouping of the properties
+   * in a property editor
+   * for example different editor tabs
+   */
+  groupId?: string;
+
+  /**
+   * allows custom sorting of the properties
+   * if 'order' is not provided, the props
+   * will be sorted by the order/key of the object (unreliable)
+   */
+  order?: number;
+}
+
+export interface StoryPropertyText extends StoryPropertyBase<string> {
+  type: PropertyTypes.TEXT;
+
+  /**
+   * placeholder for empty properties
+   * either undefined initial value
+   * or user clears the field
+   */
+  placeholder?: string;
+
+  /**
+   * number of rows in a TextArea field
+   * by default, only 1 row = means a Input field
+   */
+  maxRows?: number;
+}
+
+export interface StoryPropertyBoolean extends StoryPropertyBase<boolean> {
+  type: PropertyTypes.BOOLEAN;
+}
+
+export interface StoryPropertyColor extends StoryPropertyBase<string> {
+  type: PropertyTypes.COLOR;
+}
+
+export interface StoryPropertyDate extends StoryPropertyBase<Date> {
+  type: PropertyTypes.DATE;
+  /**
+   * whether to display a date picker (calendar).
+   * default: true
+   */
+  datePicker?: boolean;
+
+  /**
+   * whether to display a time picker (calendar).
+   * default: true
+   */
+
+  timePicker?: boolean;
+}
+
+export interface StoryPropertyFiles extends StoryPropertyBase<string[]> {
+  type: PropertyTypes.FILES;
+  /**
+   * type of files to accept user to open
+   * ex 'image/*',
+   */
+  accept?: string;
+}
+
+export interface StoryPropertyArray extends StoryPropertyBase<string[]> {
+  type: PropertyTypes.ARRAY;
+  /**
+   * the array items separator, by default comma
+   */
+  separator?: string;
+}
+
+export interface StoryPropertyObject extends StoryPropertyBase<object> {
+  type: PropertyTypes.OBJECT;
+}
+
+export interface StoryPropertyButton extends StoryPropertyBase<void> {
+  type: PropertyTypes.BUTTON;
+
+  /**
+   * for button type fields, an onClick handler
+   */
+  onClick?: (prop: StoryPropertyButton) => void;
+}
+
+export type OptionsValueType =
+  | string
+  | number
+  | string[]
+  | number[]
+  | { label: string; value: any };
+
+/**
+ * value/label pairs or array of OptionsValueType
+ */
+export type OptionsListType = { [key: string]: string } | OptionsValueType[];
```

Comment:
Am I right here that
```
{ label: string; value: any }[]
```
and
```
{ [key: string]: string }
```
might be a bit confusing. Is it possible to make it generic so people can use them with the type they want?

Reply:
I mean it's might be useful to define options in kind of this way:
```
export type OptionsValueType<T> =
   | T
   | string
   | number
   | string[]
   | number[]
   | { label: string; value: T };

export type OptionsListType<T> = { [key: string]: T } | OptionsValueType<T>[];
```

Reply:
Thanks for the feedback, yes its all possible and best to get it accepted and then make other PRs with revisions. The whole csf lib is a early version like 0.01 and a better project management practice would have been to get this started by merging the PR and then make separate issues/PRs for all suggestions and fixes.

As it stands I am pretty much over it, and will probably cancel the PR :(

Reply:
I think that's valuable to have @BeInLife! I added it.

All this looks good great to me!